### ISSUE TITLE
Claims not being added to the JWT token.

### ISSUE THREAD

**Issue by eestein** [aspnet-contrib/AspNet.Security.OpenIdConnect.Server#514]
Could anyone help me out here? I can't see what I'm doing wrong... The claims I add while generating the token are not being "deserialized?" by .net.

Controller:

```csharp
[HttpGet("user")]
public HttpResponseMessage GetUser()
{
    return new HttpResponseMessage(HttpStatusCode.OK)
    {
        Content = new StringContent("I'm logged!")
    };
}
```

Token creation:

```csharp
var identity = new ClaimsIdentity(
    context.Scheme.Name,
    OpenIdConnectConstants.Claims.Name,
    OpenIdConnectConstants.Claims.Role
);

identity.AddClaim(OpenIdConnectConstants.Claims.Subject, user.Username); // added claim
identity.AddClaim(OpenIdConnectConstants.Claims.Email, user.Email); // added claim

var ticket = new AuthenticationTicket(
    new ClaimsPrincipal(identity),
    new AuthenticationProperties(),
    context.Scheme.Name
);

ticket.SetScopes(new[]
{
    OpenIdConnectConstants.Scopes.OpenId,
    OpenIdConnectConstants.Scopes.Email,
    OpenIdConnectConstants.Scopes.Profile,
    OpenIdConnectConstants.Scopes.OfflineAccess
}.Intersect(context.Request.GetScopes()));

ticket.SetResources(Constants.Security.Token.Audience);

context.Validate(ticket);
```

The claims are being added, but when I try to access a route with that token `HttpContext.User.Claims` does not contain them:

![screen shot 2018-04-16 at 17 15 29](https://user-images.githubusercontent.com/2049255/38833087-e83f2506-419a-11e8-9ccc-051762098b85.jpg)

Am I missing a step here?

Thanks!

**Comment by eestein** [comment]
Apparently I was missing the `SetDestinations`:

```csharp
identity.AddClaim(
    new Claim(OpenIdConnectConstants.Claims.Subject, user.Username)
        .SetDestinations(OpenIdConnectConstants.Destinations.AccessToken)
);
identity.AddClaim(
    new Claim(OpenIdConnectConstants.Claims.Email, user.Email)
        .SetDestinations(OpenIdConnectConstants.Destinations.AccessToken)
);
```

I have two more questions though:

1 - Is this the correct way of doing it?
2 - Why did the claims change the name? For example:

Under `HttpContext.User.Claims` I have `http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress` instead of `email`.

Because of that I'm unable to retrieve data like this:

```csharp
var subject = HttpContext.User.GetClaim(OpenIdConnectConstants.Claims.Subject);
var email = HttpContext.User.GetClaim(OpenIdConnectConstants.Claims.Email);
```

Thanks!

**Comment by kevinchalet** [comment]
> Is this the correct way of doing it?

Yep.

> Why did the claims change the name? For example:

Because of `JwtBearerHandler`s built-in claims mapping feature: https://leastprivilege.com/2016/08/21/why-does-my-authorize-attribute-not-work/

**Comment by eestein** [comment]
Awesome, read it all, really enlightening. Thanks man!

**Comment by eestein** [comment]
@PinpointTownes last night after reading your post I tried I few things with no success. It was clear to me the problem was not related to your library, since jwt.io could correctly decode the token and displayed the correct claims. But after accessing a route I still did not receive the correct claims, they had been mapped. It was only after a lot of searching I found out the problem... I was missing this:

```csharp
.AddJwtBearer(options =>
{
    options.RequireHttpsMetadata = Constants.Server.Environment == EnvironmentType.Production;
    options.Audience = Constants.Security.Token.Audience;
    options.TokenValidationParameters = JwtToken.ValidationParameters;

    var validator = options.SecurityTokenValidators.OfType<JwtSecurityTokenHandler>().SingleOrDefault();

    if (validator != null)
    {
        validator.InboundClaimTypeMap = new Dictionary<string, string>();
        validator.OutboundClaimTypeMap = new Dictionary<string, string>();
    }
})
```

It was only after removing the map from the validator it started working.

I only have one final question... is this the right place to do it? Would say it should be done somewhere else?

Sorry to be asking so many questions, but honestly, you seem to be the most experienced person on the subject on the internet, I'm fairly new to this approach and most of what I've learned came from digging through your lib's code.

Thanks again.
